source/common/common/singleton.h

```diff
@@ -12,9 +14,30 @@ template <class T> class ConstSingleton {
    * @return const T& a reference to the singleton for class T.
    */
   static const T& get() {
-    static T* instance = new T();
-    return *instance;
+    absl::call_once(create_once_, &ConstSingleton<T>::Create);
+    return *instance_;
+  }
+
+protected:
+  template <typename A> friend class TestThreadsafeSingletonInjector;
+
+  static void Create() { instance_ = new T(); }
+
+  static absl::once_flag create_once_;
+  static T* instance_;
+};
+
+/* Mutable singleton.  All functions in the singleton class *must* be threadsafe.*/
+template <class T> class ThreadSafeSingleton : public ConstSingleton<T> {
```

Comment:
It's a bit confusing to have the mutable thread safe singleton inherit from `ConstSingleton`. Is there a huge benefit here? Do we ever even want to mock out the stuff in `ConstSingleton` (it's mostly just literals/constant strings etc.)?